Ghidra/Processors/8048/data/languages/8048.slaspec

```diff
@@ -0,0 +1,349 @@
+attach names bs      [ RB0 RB1 ];
+attach names pp      [ BUS P1 P2 _ ];
+attach names xpp     [ P4 P5 P6 P7 ];
+
+
+################################################################
+# Psuedo Instructions
+################################################################
+
+define pcodeop nop;
+define pcodeop enableExtInt;
+define pcodeop enableTCntInt;
+define pcodeop enableClockOutput;
+define pcodeop disableExtInt;
+define pcodeop disableTCntInt;
+define pcodeop startTimer;
+define pcodeop startEventCounter;
+define pcodeop stopTimerAndEventCounter;
+define pcodeop setTmr;
+define pcodeop getTmr;
+define pcodeop getT0;
+define pcodeop getT1;
+define pcodeop getTF;
+define pcodeop getExtInt;
+define pcodeop readPort;
+define pcodeop writePort;
+define pcodeop setBank;
+
+
+################################################################
+# Macros
+################################################################
+
+macro getPSW(reg) {
+  local tmp:1 = 0;
+  tmp[7,1] = C;
+  tmp[6,1] = AC;
+  tmp[5,1] = F0;
+  tmp[4,1] = BS;
+  tmp[3,1] = 1;
+  tmp[0,3] = (SP>>1)&7;
+  reg = tmp;
+}
+
+macro setPSW(reg) {
+  local tmp:1 = reg;
+  C  = tmp[7,1];
+  AC = tmp[6,1];
+  F0 = tmp[5,1];
+  BS = tmp[4,1];
+  SP = 2*tmp[0,3] + 8;
+}
+
+macro savePSWtoPC(pc) {
+  pc[15,1] = C;
+  pc[14,1] = AC;
+  pc[13,1] = F0;
+  pc[12,1] = BS;
+}
+
+macro restorePSWfromPC(pc) {
+  C  = pc[15,1];
+  AC = pc[14,1];
+  F0 = pc[13,1];
+  BS = pc[12,1];
+}
+
+macro push(v) {
+  *[INTMEM]:2 SP = v;
+  SP = SP + 2;
+}
+
+macro pop(v) {
+  SP = SP - 2;
+  v = *[INTMEM]:2 SP;
+}
+
+macro popPC(pc) {
+  pop(pc);
+  pc = pc & 0xfff;
+}
+
+macro popPCandPSW(pc) {
+  pop(pc);
+  restorePSWfromPC(pc);
+  pc = pc & 0xfff;
+}
+
+macro funcall(target) {
+  ret:2 = inst_next; 
+  savePSWtoPC(ret);
+  push(ret);
+  call target;
+}
+
+macro add(dest, op1, op2, cy_in) {
+  local result:1 = op1 + op2 + cy_in;
+  local half_result:1 = (op1 & 0xf) + (op2 & 0xf) + cy_in;
+  C  = carry(op1, op2) || carry(op1+op2, cy_in);
+  AC = (half_result > 0xf);
+  dest = result;
+}
+
+macro da(reg) {
+  local tmp:1 = reg;
+  local low:1 = 6*(AC || (tmp&0xf) > 9);
+  local cy1:1 = C || carry(tmp, low);
+  tmp = tmp + low;
+  local high:1 = 0x60*(cy1 || tmp > 0x99);
+  C = C || carry(tmp, high);
+  tmp = tmp + high;
+  reg = tmp;
+}
+
+macro rotc(cy, acc) {
+  local tmp:1 = cy;
+  A = acc;
+  C = tmp;
+}
+
+macro xch(node1, node2) {
+  local tmp:1 = node1;
+  node1 = node2;
+  node2 = tmp;
+}
+
+@ifdef SINGLE_REGISTER_BANK
+macro regbank(r) { }
+macro setbank(bs) {
+  BS = bs;
+  local tmp:1 = bs;
+  setBank(tmp);
+}
+@else
+macro regbank(r) {
+  r = r + BS*0x18;
+}
+macro setbank(bs) {
+  BS = bs;
+}
+@endif
+
+
+################################################################
+
+Psw: "PSW" is epsilon { }
+ExtInt: "I" is epsilon { }
+TCntInt: "TCNTI" is epsilon { }
+Clk: "CLK" is epsilon { }
+Tmr: "T" is epsilon { }
+Cnt: "CNT" is epsilon { }
+TmrCnt: "TCNT" is epsilon { }
+
+Rn: rn is rn & rnfill=1  {
```

Comment:
The Sleigh editor had an issue displaying this, so I re-factored it.  I think that it would be easier just to always use names for the registers, and use mapped registers with BS set to a default.

@ifdef INTERNAL_REGISTERS
Rn: rn  is rn & rnfill=1 {
  export rn;
}

Rind: @ri  is ri & rifill=0 {
  export ri;
}

@else
Rn: rn  is rn & rnfill=1 {
        local ptr:1 = rn;
        regbank(ptr);
        export *[INTMEM]:1 ptr;
}

Rind: @ri  is ri & rifill=0 {
        local ptr:1 = ri;
        regbank(ptr);
        export *[INTMEM]:1 ptr;
}

@endif